graphql/src/execution/execution.rs

```diff
@@ -263,24 +271,39 @@ where
         }
     }
 
-    // Execute the root selection set against the root query type
-    if data_set.items.is_empty() {
-        // Only introspection
-        execute_selection_set(&ictx, &intro_set, introspection_query_type, initial_value)
-    } else if intro_set.items.is_empty() {
-        // Only data
-        execute_selection_set(&ctx, &data_set, query_type, initial_value)
+    // If we are getting 'normal' data, prefetch it from the database
+    let mut values = if data_set.items.is_empty() {
+        BTreeMap::default()
     } else {
-        // Both introspection and data
-        let mut values = execute_selection_set_to_map(&ctx, &data_set, query_type, initial_value)?;
+        // Allow turning prefetch of as a safety valve. Once we are confident
+        // prefetching contains no more bugs, we should remove this env variable
+        let initial_data = if std::env::var_os("GRAPH_GRAPHQL_NO_PREFETCH").is_some() {
+            None
+        } else {
+            ctx.resolver.prefetch(&ctx, &data_set)?
+        };
+        let values = execute_selection_set_to_map(&ctx, &data_set, query_type, &initial_data)?;
+        if ctx.mode == ExecutionMode::Verify {
+            let single_values = execute_selection_set_to_map(&ctx, &data_set, query_type, &None)?;
```

Comment:
I am not attached to the name, but at a loss for a better one - any suggestions? ('single' here refers to the fact that we used to resolve single objects, rather than collections of them)